File: 자료구조/Python/백준-데큐-17413.py
from sys import stdin
import queue

def main():
    sentence = stdin.readline().strip()

    stack = []
    queue = []

    for alpha in sentence:

        if alpha == '<':
            while stack:
                print(stack.pop(), end="")
            queue.append(alpha)
            continue
        if queue:
            if alpha == '>':
                while queue:
                    print(queue.pop(0), end="")
                print(alpha, end="")
            else:
                queue.append(alpha)
        else:
            if alpha == ' ':
                while stack:
                    print(stack.pop(), end="")
                print(alpha, end="")
            elif alpha == '\n':
                continue
            else:
                stack.append(alpha)

    while queue:
        print(queue.pop(0), end="")

    while stack:
        print(stack.pop(), end="")


if __name__ == "__main__":
    main()

File: 자료구조/C++/백준-데큐-17413.cpp
# include <iostream>
# include <string>
# include <stack>
# include <queue>

using namespace std;

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    string sentence;
    getline(cin, sentence);

    stack<char> st;
    queue<char> q;
    bool open = false;


    for (char c : sentence)
    {
        if (c == '<')
        {
            while (!st.empty())
            {
                cout << st.top();
                st.pop();
            }
            q.push(c);
            open = 1;
            continue;
        }
        else if (c == '>') 
        {
            q.push(c);
            while (!q.empty())
            {
                cout << q.front();
                q.pop();
            }
            open = 0;
            continue;
        }
        if (open) 
        {
            q.push(c);
        }
        else 
        {
            if (c == ' ')
            {
                while (!st.empty())
                {
                    cout << st.top();
                    st.pop();
                }
                cout << ' ';
            }
            else if (isalnum(c))
            {
                st.push(c);
            }
        }
        
        
    }
    while (!st.empty())
    {
        cout << st.top();
        st.pop();
    }
    while (!q.empty())
    {
        cout << q.front();
        q.pop();
    }

    return 0;
}